cirq/google/engine/engine_client.py

```diff
@@ -601,3 +609,188 @@ def get_current_calibration(self, project_id: str, processor_id: str
+            lambda: self.grpc_client.list_quantum_reservations(
+                self._processor_name_from_ids(project_id, processor_id),
+                filter_=filter_str))
+
+        return list(response)
+
+    def update_reservation(self,
+                           project_id: str,
+                           processor_id: str,
+                           reservation_id: str,
+                           start: Optional[datetime.datetime] = None,
+                           end: Optional[datetime.datetime] = None,
+                           whitelisted_users: Optional[List[str]] = None):
+        """Updates a quantum reservation.
+
+        This will update a quantum reservation's starting time, ending time,
+        and list of whitelisted users.  If any field is not filled, it will
+        not be updated.
+
+        Params:
+            project_id: A project_id of the parent Google Cloud Project.
+            processor_id: The processor unique identifier.
+            reservation_id: Unique ID of the reservation in the parent project,
+            start: the new starting time of the reservation as a datetime object
+            end: the new ending time of the reservation as a datetime object
+            whitelisted_users: a list of emails that can use the reservation.
+        """
+        name = self._reservation_name_from_ids(
+            project_id, processor_id, reservation_id) if reservation_id else ''
+
+        reservation = qtypes.QuantumReservation(name=name,)
+        paths = []
+        if start:
+            reservation.start_time.seconds = int(start.timestamp())
+            paths.append('start_time')
+        if end:
+            reservation.end_time.seconds = int(end.timestamp())
+            paths.append('end_time')
+        if whitelisted_users:
+            reservation.whitelisted_users.extend(whitelisted_users)
+            paths.append('whitelisted_users')
+
+        return self._make_request(
+            lambda: self.grpc_client.update_quantum_reservation(
+                name=name,
+                quantum_reservation=reservation,
+                update_mask=qtypes.field_mask_pb2.FieldMask(paths=paths)))
+
+    def list_time_slots(self,
+                        project_id: str,
+                        processor_id: str,
+                        filter_str: str = '') -> List[qtypes.QuantumTimeSlot]:
+        """Returns a list of quantum time slots on a processor.
+
+        Params:
+            project_id: A project_id of the parent Google Cloud Project.
+            processor_id: The processor unique identifier.
+            filter:  A string expression for filtering the quantum
+                time slots returned by the list command. The fields
+                eligible for filtering are `start_time`, `end_time`.
+
+        Returns:
+            A list of QuantumTimeSlot objects.
+        """
+        response = self._make_request(
+            lambda: self.grpc_client.list_quantum_time_slots(
```

Comment:
This method is paginated and won't necessarily return all values in a single request. Paged responses should be compiled on the client side.

Reply:
I think this is already taken care of in the gapic client:

https://github.com/quantumlib/Cirq/blob/master/cirq/google/engine/client/quantum_v1alpha1/gapic/quantum_engine_service_client.py#L1887

Reply:
Well that's cool.